File: src/main/java/io/github/jolkert/discordbotlink/jda/command/WhitelistCommand.java
package io.github.jolkert.discordbotlink.jda.command;

import com.mojang.authlib.GameProfile;
import io.github.jolkert.discordbotlink.data.UserData;
import io.github.jolkert.discordbotlink.DiscordBotLink;
import net.dv8tion.jda.api.EmbedBuilder;
import net.dv8tion.jda.api.entities.MessageEmbed;
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent;
import net.dv8tion.jda.api.events.message.MessageReceivedEvent;
import net.minecraft.server.ServerConfigHandler;
import net.minecraft.server.Whitelist;
import net.minecraft.server.WhitelistEntry;
import org.jetbrains.annotations.NotNull;

import java.util.Objects;
import java.util.UUID;

public class WhitelistCommand extends DiscordCommand
{
	public static final String NAME = "whitelist";

	public WhitelistCommand()
	{
		this.name = NAME;
		this.aliases = new String[]{"wl"};
	}
	// TODO: maybe make this accept UUIDs? idk
	@Override
	public void execute(String[] args, MessageReceivedEvent event)
	{
		if (args.length < 2 || args[1].isBlank())
			return;

		MessageEmbed embed = addToWhitelist(args[1], CommandContext.of(event));
		event.getChannel().sendMessageEmbeds(embed).queue();
	}
	@Override
	public void executeSlashCommand(@NotNull SlashCommandInteractionEvent event)
	{
		event.deferReply().queue();
		MessageEmbed embed = addToWhitelist(Objects.requireNonNull(event.getOption("username")).getAsString(), CommandContext.of(event));
		event.getHook().sendMessageEmbeds(embed).queue();
	}

	private MessageEmbed addToWhitelist(String username, CommandContext context)
	{
		Whitelist whitelist = DiscordBotLink.Server.getPlayerManager().getWhitelist();
		
		MessageEmbed embed;
		DiscordBotLink.Logger.info("Adding [" + username + "] to the whitelist");
		UUID uuid = ServerConfigHandler.getPlayerUuidByName(DiscordBotLink.Server, username);
		if (uuid == null)
		{
			embed = new EmbedBuilder()
					.setTitle("User not found!")
					.setDescription("User **<" + username + ">** could not be found!")
					.setColor(0xff00000)
					.build();
		}
		else
		{
			GameProfile userProfile = new GameProfile(uuid, username);
			if (whitelist.isAllowed(userProfile))
			{
				embed = new EmbedBuilder().setTitle("Cannot add to whitelist!")
						.setDescription("User **<" + username + ">** is already on the whitelist!")
						.setColor(0xff0000)
						.build();
			}
			else
			{
				WhitelistEntry entry = new WhitelistEntry(userProfile);
				UserData oldData = DiscordBotLink.Bot.getUserData().getUser(context.author());
				if (oldData != null)
				{
					oldData = new UserData(oldData);
					
					UserData user = DiscordBotLink.Bot.getUserData().getUser(context.author());
					GameProfile toRemove = new GameProfile(oldData.getMinecraftData().getUuid(), oldData.getMinecraftData().getName());
					DiscordBotLink.Bot.getUserData().updateMinecraftData(context.author(), userProfile);
					whitelist.remove(toRemove);
				}
				else
				{
					UserData userData = new UserData(userProfile, context.author());
					DiscordBotLink.Bot.getUserData().addUser(userData);
				}
				
				whitelist.add(entry);
				embed = new EmbedBuilder()
						.setTitle("User added to whitelist!")
						.setDescription("User **<" + username	+ ">** added to the whitelist, and linked to " + context.author().getAsMention() +
								(oldData != null ? "\nReplaced **<" + oldData.getMinecraftData().getName() + ">**" : ""))
						.setColor(0x009900)
						.build();
			}
			
		}
		
		return embed;
	}
}

File: src/main/kotlin/io/github/jolkert/discordbotlink/jda/command/WhitelistCommand.kt
package io.github.jolkert.discordbotlink.jda.command

import com.mojang.authlib.GameProfile
import io.github.jolkert.discordbotlink.DiscordBotLink
import io.github.jolkert.discordbotlink.DiscordBotLink.LOGGER
import io.github.jolkert.discordbotlink.jda.data.CommandContext
import io.github.jolkert.discordbotlink.jda.data.UserData
import net.dv8tion.jda.api.EmbedBuilder
import net.dv8tion.jda.api.entities.MessageEmbed
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent
import net.dv8tion.jda.api.events.message.MessageReceivedEvent
import net.dv8tion.jda.api.interactions.commands.OptionType
import net.dv8tion.jda.api.interactions.commands.build.Commands
import net.minecraft.server.ServerConfigHandler
import net.minecraft.server.WhitelistEntry

class WhitelistCommand : Command()
{
	override val name = "whitelist"
	override val aliases = arrayOf("wl")
	override val slashCommandData =
		Commands.slash("whitelist", "Add a Minecraft user to the whitelist")
			.addOption(
				OptionType.STRING,
				"username",
				"Your minecraft username. You can only add one Minecraft account per Discord account.",
				true
			)
			.setGuildOnly(true)

	override fun executeTextCommand(event: MessageReceivedEvent, vararg args: String)
	{
		if (args.size < 2 || args[1].isBlank())
			return

		val embed = addToWhitelist(args[1], CommandContext.of(event)).toEmbed()
		event.channel.sendMessageEmbeds(embed).queue()
	}

	override fun executeSlashCommand(event: SlashCommandInteractionEvent)
	{
		val user = event.getOption("username")!!.asString

		event.deferReply().queue()
		event.hook.sendMessageEmbeds(addToWhitelist(user, CommandContext.of(event)).toEmbed()).queue()
	}


	class WhitelistAddResult(val type: ResultType, val username: String, val replacedUser: String? = null)
	{


		enum class ResultType
		{
			SUCCESS,
			USER_ALREADY_WHITELISTED,
			USER_NOT_FOUND
		}
	}



	companion object
	{
		fun addToWhitelist(username: String, context: CommandContext): WhitelistAddResult
		{
			val whitelist = DiscordBotLink.Server.playerManager.whitelist

			LOGGER.info("Attempting to add <$username> to the whitelist")
			val uuid = ServerConfigHandler.getPlayerUuidByName(DiscordBotLink.Server, username)
				?: return WhitelistAddResult(WhitelistAddResult.ResultType.USER_NOT_FOUND, username)

			val userProfile = GameProfile(uuid, username)
			if (whitelist.isAllowed(userProfile))
				return WhitelistAddResult(WhitelistAddResult.ResultType.USER_ALREADY_WHITELISTED, username)

			val oldData = DiscordBotLink.users[context.author.id]
			if (oldData == null)
				DiscordBotLink.users.add(UserData(userProfile, context.authorAsMember!!))
			else
			{
				whitelist.remove(oldData.minecraftData.toGameProfile())
				DiscordBotLink.users[oldData.uuid] =
					UserData(oldData, minecraftData = UserData.MinecraftUserData(userProfile))
			}

			whitelist.add(WhitelistEntry(userProfile))
			return WhitelistAddResult(WhitelistAddResult.ResultType.SUCCESS, username, oldData?.minecraftData?.username)
		}

		fun WhitelistAddResult.toEmbed(): MessageEmbed
		{
			return EmbedBuilder().apply {
				(when (type)
				{
					WhitelistAddResult.ResultType.SUCCESS -> setTitle("Success!")
						.setDescription("Successfully added user **<$username>**".let {
							if (replacedUser == null)
								return@let it
							else
								return@let "$it\nReplaced *<$replacedUser>*"
						})
						.setColor(0x00ff00)

					WhitelistAddResult.ResultType.USER_ALREADY_WHITELISTED -> setTitle("Already Whitelisted!")
						.setDescription("Sorry, **<$username>** is already on the whitelist!")
						.setColor(0xff0000)

					WhitelistAddResult.ResultType.USER_NOT_FOUND -> setTitle("User Not Found!")
						.setDescription("No Minecraft account with the name **<$username>** was found!\nDouble check the spelling of your username!")
						.setColor(0xff0000)
				})
			}.build()
		}
	}
}